---
title: "r4rnaseq-workshop"
author: "Rayna M Harris"
output: md_document
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, cache = TRUE,
                      fig.path = "./images/",
                      appendix = TRUE)
```

# Introduction to R for RNA-Seq Workshop

---

_Even though you can [![hackmd-github-sync-badge](https://hackmd.io/2ArmQGwGT0uUyL5Ehqy0hQ/badge)](https://hackmd.io/2ArmQGwGT0uUyL5Ehqy0hQ), please make changes by editing [this .Rmd file](https://github.com/nih-cfde/training-rstudio-binder/blob/data/GTEx/r4rnaseq-workshop.Rmd)._

---

**Description**

This free two-hour workshop will provide an overview of the R programming language and the user-friendly RStudio environment for exploratory RNA-sequencing analysis. You will be introduced to R syntax, variables, functions, packages, and data structures. You will learn the basics of data wrangling including importing data from files, sub-setting, joining, filtering, and more. You will gain hands-on practice calculating and visualizing differential gene expression using popular open-source packages and public RNA-Seq data from the Gene Expression Tissue Project (GTEx). _Please fill out the pre- and post-workshop surveys to help us keep these workshops free._

**When:** Wednesday, March 23, 10 am - 12 pm PT  
**Where:** [Zoom](https://zoom.us/j/7575820324?pwd=d2UyMEhYZGNiV3kyUFpUL1EwQmthQT09)  
**Instructors:** Dr. Rayna Harris  
**Helpers:** Dr. Amanda Charbonneau and Dr. Jessica Lumian

Click [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/nih-cfde/training-rstudio-binder/data?urlpath=rstudio) to generate a computing environment for this workshop.


### Overview

```{asis, comment = "", appendix = F}

:::info
#### Learning Objectives 

In this workshop, you will learn how to use R and RStudio to import, tidy, transform, and visualize data structures commonly associated with RNA-sequencing experiments. Specifically, you will: 

* Explore public RNA-Seq data from the Gene Expression Tissue Project
* Select variables and observations that are relevant to research questions 
* Rename variables and experimental factors for data joining and plotting
* Visualize raw and summarized data using bar graphs, scatter plots, and box plots 

:::

```

[TOC]

## Introduction 

The book [“R for Data Science”](https://r4ds.had.co.nz/index.html) provides an excellent framework for using data science to turn raw data into understanding, insight, and knowledge. We will use this framework as an outline for this workshop.

**R** is a statistical computing and data visualization programming language. **RStudio** is an integrated development environment, or IDE, for R programming. R and RStudio work on Mac, Linux, and Windows operating systems. The RStudio layout displays lots of useful information and allows us to run R code from a script and view and save outputs all from one interface.

When you start RStudio, you’ll see two key regions in the interface: the console and the output. When working in R, you can type directly into the console, or you can type into a script. Saving commands in a script will make it easier to reproduce. You will learn more as we go along! 

![](https://hackmd.io/_uploads/SkkxxSHeq.png =300x)
![](https://hackmd.io/_uploads/H1a8-HHx5.png =300x)

For today's lesson, we will focus on data from the [Gene Expression Tissue (GTEx) Project](https://commonfund.nih.gov/gtex). The GTEx is an ongoing effort to build a comprehensive public resource to study tissue-specific gene expression and regulation. Samples were collected from 54 non-diseased tissue sites across nearly 1000 individuals, primarily for molecular assays including WGS, WES, and RNA-Seq. 

By the end of today’s workshop, you create tables and plots like the
ones below that give an overview of the samples collected
and the variables that can be used for differential gene expression
analysis. More importantly, you will learn the basics of importing, tidying, transforming, and visualizing data, which are the key component of an R workflow. 

### Some motivating questions

* How many RNA-sequencing samples are in the GTEx project? 
* What is the age and sex of each donor?
* What was the cause of death?
* What is the effect of age on gene expression in the heart?
* Do you have enough samples to test the effects of sex, age, hardy scale, and their interactions for all tissues? 
* How do I combine, clean, modify, separate, etc. data sets and variables?
* How is my gene of interest affected by age in the heart and muscle?

![](https://hackmd.io/_uploads/SJSIB76b9.png =300x) ![](https://hackmd.io/_uploads/Sk6IrQaZc.png =300x)

![](https://hackmd.io/_uploads/r1mwBQaW5.png =300x) ![](https://hackmd.io/_uploads/HJ5BrXp-c.png =300x)

### Getting Started

1. Click the [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/nih-cfde/training-rstudio-binder/data?urlpath=rstudio) button to generate a computing environment for this workshop.
2. Navigate to the GTEX folder.
3. Click `GTEx.Rproj` and click "Yes" to open up an Rproject. This will set the working directory to `~/GTEx/`.
4. Open the `r4rnaseq-workshop.R` file which contains all the commands for today's workshop. This contains all the commands we will build today. This is your reference.
5. Open a new R Script by clicking **File > New File > R Script**. You will type most commands for today's lesson here and click "Run" to send them to the console. 


### R packages

[**Tidyverse**](https://www.tidyverse.org/) is a collection of R packages that include functions, data, and documentation that provide more tools and capabilities when using R. There are many packages for R, but we are using this set because the packages are designed to work well together -and they are especially useful (and popular!) for doing data science. 

You can install the complete tidyverse with a single line of code: `install.packages("tidyverse")`, or you can install packages individually (e.g. `install.packages("ggplot2")`). It is a good idea to "comment out" this line of code by adding a `#` at the beginning so that you don't re-install the package every time you run the script. For this workshop, the packages listed in the `.binder/environment.yml` file were pre-installed with Conda. For some reason, the tidyverse package doesn't always install properly, so we installed each package individually.

```{r  eval=FALSE}
#install.packages("tidyverse")
#install.packages("ggplot2")
#install.packages("tidyr")
```

After installing packages, we need to load the functions and tools we want to use from the package with the `library()` command. Let's load the following packages: `ggplot2, tidyr, dplyr, readr, and tibble` by copying and pasting or typing these commands into your new script.

```{r, message=FALSE, appendix = TRUE}
library(ggplot2)
library(tidyr)
library(dplyr)
library(readr)
library(tibble)
```

```{r, otherpackages, echo = F, message = F, warning=F , appendix = TRUE}
library(cowplot)
library(scales)
library(forcats)
library(stringr)
```


```{asis, comment = "" , appendix = F}

:::warning

#### Challenge

We will also use `cowplot` and  `scales` to make pretty visualizations, `forcats.` for working with factors, and `stringr` for parsing text. What commands do you need to add to your script to load these packages?

:::spoiler 

  `library(cowplot)`  
  `library(scales)`  
  `library(forcats)`  
  `library(stringr)`  
:::

```

<!---
[Bioconductor](https://www.bioconductor.org) is an open-source software project developed for the analysis and comprehension of high-throughput data in genomics and molecular biology. The project aims to enable interdisciplinary research, collaboration, and rapid development of scientific software. It is based on the statistical programming language R. We will not be using these packages in this class, but they are worth getting to know. This is how you would install and load 3 different Bioconductor packages. 

```{r, message=FALSE, eval=F}

#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install()

#BiocManager::install(c("recount3", "biomaRt", "DESeq2"))

#library(recount3)
#library(biomaRt)
#library(DESeq2)
```
--->

You can also navigate to "Packages" Tab in the bottom right pane of RStudio to view a list of available packages. Packages with a checked box next to them have been successfully loaded. You can click a box to load installed packages. Clicking the "Help" Tab will provide a quick description of the package and its functions.



```{asis, comment = "", appendix = F}

:::success
#### Key functions

| Function | Description |
| --- | --- |
| `install.packages()` | An R function to install packages |
| `library()` | The command used to load installed packages |
:::

```


## Import

Data can be imported using packages from base R or the tidyverse. What are some differences between the data objects imported by base R functions such as `read.csv()` and Tidyverse functions such as `read_csv()`? To begin with, `read.csv()` replaces spaces and dashes periods in column names, and it also preserves row.names. On the other hand, `read_csv()` preserves spaces and dashes in column names but drops row.names. For this workshop, we will use `read_csv()`, which means we may have to replace dashes with periods so that our sample names in all objects with sample name information.

Today, I will show you how to import the following files:

1. data/GTExPortal.csv
1. data/GTExHeart_20-29_vs_70-79.tsv
1. data/colData.HEART.csv
1. data/countData.HEART.csv.gz


Then, you will practice on your own using the following files:

1. data/GTExMuscle_20-29_vs_70-79.tsv
1. data/colData.MUSCLE.csv
1. data/countData.MUSCLE.csv.gz


The `GTExPortal.csv` file in `./data/` contains information about all the samples in the GTEx portal. Let's import this file using `read.csv()`. 

```{r import-1, appendix = TRUE}

samples <- read.csv("./data/GTExPortal.csv")

```
After importing a file, there are multiple was to view the data. `head()` to view the first few lines of each file. `names()` will print just the column names. 

```{r, appendix = TRUE}

head(samples)
names(samples)

```

Very large tabular files are often saved as .tsv files. These can be imported with `read.table()` or  `read_tsv()`. You can also specify the tab delimiter as well as the row and column names. You can import files using the default parameters or you can change them. Because the first column in the .tsv files does not have a row name, by default, `read.table()`, imports the first column as the row.names. When `sep = "\t", header = TRUE` is specified, the fist column is imported as column one and given the column name  `X`.

```{r import-2, appendix = TRUE}

# a data frame with row.names
results <- read.table("./data/GTEx_Heart_20-29_vs_30-39.tsv")
head(results)

# a data frame without row.names
results2 <- read.table("./data/GTEx_Heart_20-29_vs_30-39.tsv",  sep = "\t", header = TRUE )
head(results2)

```

I find it helpful to import a table of gene names and symbols that can be merged with other tables with gene information or searched for useful information. This table of gene names was downloaded from <https://www.genenames.org/download/custom/>. 

In this case, we use `fill = T` to fill missing fields with a NULL value.

```{r import-3, appendix = TRUE}

genes <- read.table("./data/genes.txt", sep = "\t",  header = T, fill = T)
head(genes)
dim(genes)

```


```{asis, comment = "", appendix = F}

:::info Using the Terminal to uncompress files

Very large data files, such as files with RNA-Seq counts are often compressed before they are shared. To uncompress a file, click on the Terminal tab and run the following command.

  gunzip -k ./data/countData.HEART.csv.gz

:::

```

```{bash, echo = F, appendix = F}
rm ./data/countData.HEART.csv
gunzip -k ./data/countData.HEART.csv.gz
```


```{bash, echo = F, appendix = TRUE}

# Open the Terminal and type the command (ater the $) to uznip
# $ gunzip -k ./data/countData.HEART.csv.gz

```



Once that file is uncompressed, it can be imported. Count files can be very long and wide, so it is a good idea to only view the first (or last) few rows and columns. Typically, a gene identifier (like an ensemble id) will be used as the row names. We can use `dim` to see how many rows and columns are in the file. 




```{r import-4, appendix = TRUE}

counts <- read.csv("./data/countData.Muscle.csv", 
                   header = TRUE, row.names = 1)
dim(counts)
head(counts)[1:5]

```

This "countData" was generated by using `recount3` as described in the file `scripts/recount3.Rmd`. It comes from a Ranged Summarized Experiment (rse) which contains quantitative information about read counts as well as quality control information and sample descriptions. The "colData" from an rse can also be obtains. This information _should_ match the information in our samples file, but there can be subtle differences in formatting Let's read the colData file for the heart samples. 



```{r import-5, appendix = TRUE}

colData <- read.csv("./data/colData.HEART.csv")
head(colData)

```


```{asis, comment = "", appendix = F}

:::warning

#### Challenge

What commands could you use to read the following files:
1. GTEx results comparing the muscles of 20-29 year old to 70-79 year olds?
1. The csv file information describing the muscle samples? 


:::spoiler

1. `read.table("./data/GTEx_MUSCLE_20-29_vs_70-79.tsv")` 
1. `read.csv("./data/colData.MUSCLE.csv")`  

:::

```

## Quick summary statistics and sample size

You have now seen a variety of options for importing files. You may use many more in your R-based RNA-seq workflow, but these basics will get you started. Let's now explore the functions `summary()`,  `length()`, `dim()`, and `count()` us to quickly summarize and compare data frames to answer the following questions.




How many transcripts were counted in the Heart tissues?  Over 63,000.

```{r, appendix = TRUE}

dim(counts)
length(row.names(counts))

```

How many genes are in this version of the human transcriptome? Over 15,000.

```{r, appendix = TRUE}

dim(genes)
length(genes$Approved.symbol)

```

How many samples are there in the GTEx portal (as of this version)? Over 25,000!

```{r, appendix = TRUE}

dim(samples)
length(samples$Tissue.Sample.ID)

```

How many samples are there per tissue? 

```{r, appendix = TRUE}

dplyr::count(samples, Tissue) 

```

How many samples are there per tissue and sex? Can we test the effect of sex on gene expression in all tissues? For many samples, yes, but not all tissues were samples from both males and females. 

```{r, appendix = TRUE}

dplyr::count(samples, Tissue, Sex) 

```

How many samples are there per sex, age, and hardy scale in the HEART tissues? Do you have enough samples to test the effects of Sex, Age, and Hardy Scale in the Heart? 

_Note: Let's use the colData data frame for this since it is specific to Heart. Later we will talk about subsetting._

```{r, appendix = TRUE}

names(colData)
dplyr::count(colData, gtex.smts, gtex.sex, gtex.age, gtex.dthhrdy) 

```


```{asis, comment = "", appendix = F}

:::warning

#### Challenge

What series commands would you use to import the `data/colData.MUSCLE.csv` and count the number of muscles samples per sex, age? 

How many female muscles samples are there from age group 30-39?

_Hint: use head() or names() after importing a file to verify the variable names._


:::spoiler

    read.csv("./data/colData.MUSCLE.csv") %>%
        dplyr::count(gtex.smts, gtex.sex, gtex.age)
    # 3 samples are in the female group age 30-39
:::

```





Finally, the `str()` and  `summary()` commands are also quite useful for summarizing every variable in a data frame. These let you know if R has imported columns properly as integers, characters or factors. 

```{r, appendix = TRUE}

str(samples)
summary(samples)

str(results2)
summary(results2)

str(counts[1:5])
summary(counts[1:5])

```


```{asis, comment = "", appendix = F}

:::success
#### Key functions for importing and quickly viewing raw and summarized data

| Function | Description |
| --- | --- |
| `read.csv()`  | A base R function for importing comma separated tabular data  |
| `read_csv()`  | A tidyR function for importing .csv files as tibbles |
| `read.table()` | A base R function for importing tabular data with any delimiter |
| `read_tsv()`  | A tidyR function for importing .tsv files as tibbles | 
| `as_tibble()` | Convert data frames to tibbles | 
| `head()` and `tail()` | Print the first or last 6 lines of an object  | 
| `dim()`  | A function that prints the dimensions of an object | 
| `length()` | Calculate the length of an object |
| `%>%` | The pipe is an operator that sends the output of one function to another function |
| `count()` | A dplyr function that counts number of samples per group |
| `str()` | A function that prints the internal structure of an object  |  
| `summary()` | A function that summarizes each variable |  

:::

```


## Tidy and Transform



### Renameing variables to join data frames

In the next section, we will join two data frames by a shared column. Both the results file and the genes file have a column with gene symbols, but they do not have the same name. 


We can use the `rename` function to rename columns. The first value is the new name and the second value is the old name. Just in case we get it wrong, let's save this as a new object.

```{r rename, appendix = TRUE}

head(results2)
head(genes)

head(results2$X)
head(genes$Approved.symbol)

results_new <- results2 %>% dplyr::rename("Approved.symbol" = "X")
head(results_new)

```

Now that the results and the genes objects both have a column called `Approved.symbol` they can be joined. The command `full_join()` will keep all rows of both objects. `left_join` will keep all the rows in the first object but will drop any rows in the second object that don't map onto the first. 

```{r join, appendix = TRUE}

results_genes <- left_join(results_new, genes, by = "Approved.symbol")
head(results_genes)

```


```{r mutate, appendix = TRUE}

head(counts)[1:5]
head(genes$Ensembl.gene.ID)

genes2 <- genes %>%
  mutate(Ensembl.gene.ID = paste(Ensembl.gene.ID, "1", sep = "."))

counts2 <- counts %>%
  mutate(Ensembl.gene.ID = row.names(.))

head(counts2$Ensembl.gene.ID)[1:5]
head(genes2$Ensembl.gene.ID)

```


Most RNA-Seq pipelines require that the counts be in a “wide” format where each sample is a column and each gene is a row. However, many R tools prefer data in the long format. I like to create a counts_long file that can be easily subset by variables or genes of interest.

For this, we must introduce the pipe, `%>%`. This symbol is used to redirect the output from standard out to another function. 



```{r, appendix = TRUE}

counts_long <- counts2 %>%
  pivot_longer(-Ensembl.gene.ID, names_to = "Tissue.Sample.ID", values_to = "counts") %>%
  inner_join(., genes2, by = "Ensembl.gene.ID") %>%
  arrange(desc(counts))
head(counts_long)

head(samples)
head(samples$Tissue.Sample.ID)
head(counts_long$Tissue.Sample.ID)

```

Yea, we have joined some data frames. Now let's do a little more cleaning and joining. 

```{r, appendix = TRUE}

head(counts_long$Tissue.Sample.ID)
head(samples$Tissue.Sample.ID)

counts_long_newname <- counts_long %>%
  separate(Tissue.Sample.ID, into = c("Tissue.Sample.ID", NULL), 
           sep =  ".SM.")

head(counts_long_newname$Tissue.Sample.ID)
head(samples$Tissue.Sample.ID)

samples_new <- samples %>%
  mutate(Tissue.Sample.ID = gsub("-", ".", Tissue.Sample.ID))
head(samples_new$Tissue.Sample.ID)


counts_long_samples <- counts_long_newname %>%
  inner_join(., samples_new, by = "Tissue.Sample.ID")
head(counts_long_samples)

```

Now you have a file with the counts every gene as well as the experimental variables in 1 data frame.


To process the counts data using the DESeq2 pipeline, we need a corresponding file where the row names are the sample id and they match the column names of the counts file. We confirm this by asking if `rownames(colData) == colnames(counts)` or by checking the dimensions of each. Using  `head()` is a good way to only print 5 rows. Using `[1:5]` is a good way to only print 5 columns.


```{r, appendix = TRUE}

head(rownames(colData) == colnames(counts))

```


<add challenge>



### Filtering, arranging and summarizing variables




```{asis, comment = "", appendix = F}

:::success
#### Key functions: Tidy and Transform

| Function | Description |
| --- | --- |
| `pivot_wider()` |  |
| `pivot_longer()` |  |
| `separate()` |  |
| `drop_na()` |  |
| `select()`|  |
| `arrange()` |  |
| `summarize()`  |  |
| `arrange()`  |  |
| `mutate()`  | | 
| `full_join()`  | | 
| `left_join()`  | | 
| `inner_join()`  | | 
:::

```


## Visualize

Now, we can use ggplot2 to show how many samples for each biological condition.

```{asis, comment = "", appendix = F}

:::success
#### The grammer of graphics

| Function | Description |
| --- | --- |
| `ggplot()`  |  |
| `aes()`  |  |
| `geom_point()`  | | 
| `geom_bar()`  | | 
| `geom_boxplot()`  | | 
| `theme()`  | | 
| `labs()`  | | 
| `scale_color_manual()`  | | 
| `cowplot()`  | | 
:::

```


## Communicate

Communication is a 2-way street. In this section, I encourage you to think not only about what you have to say but what others have to say as well. 

https://gtexportal.org/home/gene/BMP10

![](https://hackmd.io/_uploads/S1SxUwPRt.png)

### R Markdown

The workshop notes for using this repository to teach an Introduction to R for RNA-seq are crated with the file `r4rnaseq-workshop.Rmd`. 


## Functions and/or For Loops?

Not sure if I want to include either or both of these advanced beginner concepts.  


## References

- [R for Data Science by Hadley Wickham and Garrett Grolemund](https://r4ds.had.co.nz/index.html)
- [Rouillard et al. 2016. The Harmonizome: a collection of processed datasets gathered to serve and mine knowledge about genes and proteins. Database (Oxford).](http://database.oxfordjournals.org/content/2016/baw100.short)

## Additional Resources

- [RStudio cheatsheet for readr](https://raw.githubusercontent.com/rstudio/cheatsheets/master/data-import.pdf)
- [RStudio cheatsheet for dplyr](https://raw.githubusercontent.com/rstudio/cheatsheets/master/data-transformation.pdf)
- [RStudio cheatsheet for data Wrangling with dplyr](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf)
- [ggplot point shapes](http://www.sthda.com/english/wiki/ggplot2-point-shapes)
- [Angus 2019 Intro to R Lesson](https://angus.readthedocs.io/en/2019/R_Intro_Lesson.html)
- [Angus 2019 Differential Gene Expression in R Lesson](https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html)
- [Software Carpentry R Lesson](http://swcarpentry.github.io/r-novice-inflammation/)


*Note: the source document [r4rnaseq-workshop.Rmd](https://github.com/nih-cfde/training-rstudio-binder/blob/data/GTEx/r4rnaseq-workshop.Rmd) was last modified 14 March, 2022.*

---


```{r all-code, ref.label = knitr::all_labels(appendix == TRUE), eval = F, indent = ""}

```



